htlcswitch/link.go

```diff
@@ -2638,11 +2639,28 @@ func (l *channelLink) processRemoteAdds(fwdPkg *channeldb.FwdPkg,
 
 		heightNow := l.cfg.Switch.BestHeight()
 
-		fwdInfo := chanIterator.ForwardingInstructions()
+		fwdInfo, err := chanIterator.ForwardingInstructions()
+		if err != nil {
+			// If we're unable to process the onion payload, or we
+			// we received malformed TLV stream, then we should
+			// send an error back to the caller so the HTLC can be
+			// cancelled.
+			l.sendMalformedHTLCError(
+				pd.HtlcIndex, lnwire.CodeInvalidOnionVersion,
+				onionBlob[:], pd.SourceRef,
+			)
```

Comment:
Why do we actually need to use `sendMalformedHTLCError`? We have the obfuscator at this point and can encrypt the failure ourselves.

Reply:
I'd consider the onion to be malformed if we can't parse the forwarding instructions. The sender will get the same information at the other end as well. This code path will change soon in any case once we have the proper error in place.

Reply:
The difference is that for a malformed htlc error, the node's predecessor is the failure source. Our penalization in mc would be forced to blame the node pair, while the predecessor did nothing wrong.

For that reason, sending back something (temp node failure?) with ourselves as failure source is better. Even when this is not yet the final failure code.

Reply:
Modified to encrypt the error ourselves since we're able to.